### ISSUE TITLE
CSS reading-flow property

### ISSUE THREAD

**Issue by dizhang168** [w3ctag/design-reviews#978]
こんにちは TAG-さん!

I'm requesting a TAG review of the CSS reading-flow property.

The CSS Working Group has resolved to add the new reading-flow property (https://github.com/w3c/csswg-drafts/issues/8589, [spec](https://drafts.csswg.org/css-display-4/#reading-flow)) to enable focus navigation in visual order for layout items that might not be displayed in source order (such as grid, flex and masonry items).  Chromium is implementing the new property and opened a https://github.com/whatwg/html/issues/10407 on the needed HTML spec change.

  - [Explainer: HTML Spec proposal](https://github.com/dizhang168/reading-order-items)
  - [Hosted examples](https://rachelandrew.github.io/reading-order-items-examples/)
  - [Specification URL](https://drafts.csswg.org/css-display-4/#reading-flow)
  - [Tests](https://wpt.fyi/results/shadow-dom/focus-navigation/reading-flow/tentative?label=experimental&label=master&aligned)
  - User research: Multiple blog posts [link 1](https://developer.chrome.com/blog/reading-order), [link 2](https://rachelandrew.co.uk/archives/2024/05/26/masonry-and-reading-order/), [link 3](https://developer.chrome.com/blog/reading-flow-display-contents?hl=en)
  - Security and Privacy self-review²: Mostly not applicable because this is adding a new CSS property.
  - GitHub repo: https://github.com/w3c/csswg-drafts
  - Primary contacts (and their relationship to the specification):
      - Di Zhang (@dizhang168), Google, implementer in Chrome
      - Rachel Andrew (@rachelandrew), Google, spec editor
      - Tab Atkins (@tabatkins), Google, spec editor
      - Mason Freed (@mfreed7), Google
      - Ian Kilpatrick (@bfgeek), Google
  - Organization(s)/project(s) driving the specification: Google and the CSS WG
  - Key pieces of multi-implementer support:
    - Chromium comments: supportive
    - Mozilla comments: https://github.com/mozilla/standards-positions/issues/1056
    - WebKit comments: https://github.com/WebKit/standards-positions/issues/378
  - External status/issue trackers for this specification (publicly visible, e.g. Chrome Status):
    - [Chromestatus](https://chromestatus.com/feature/5113638848561152)
    - [chromium bug](https://issues.chromium.org/issues/40932006)
    - [csswg-drafts resolved issue](https://github.com/w3c/csswg-drafts/issues/8589#issuecomment-1721504153)
  - This feature should be inclusive of other API/features:
    - [Accessibility](https://github.com/w3c/aria/issues/2241)
    - [OpenUI focusgroup](https://github.com/openui/open-ui/issues/1044)
    - [Masonry](https://github.com/w3c/csswg-drafts/issues/5675)

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://www.w3.org/TR/design-principles/)
  - Relevant time constraints or deadlines: hoping to ship this property soon.
  - The group where the work on this specification is currently being done: CSS Working Group, WHATWG
  - The group where standardization of this work is intended to be done (if current group is a community group or other incubation venue): CSS Working Group, WHATWG
  - Major unresolved issues with or opposition to this specification: https://github.com/w3c/csswg-drafts/issues/9230
  - This work is being funded by: Google

**Comment by matatk** [comment]
Hi @dizhang168, thanks for your review request. I have an accessibility question: the examples that I've seen in the explainers concentrate on showing how the (keyboard) focus navigation order, and visual rendering order, would be kept in synch when using `reading-flow` (which will be great, especially for masonry layout).

The examples I've seen all use interactive elements (specifically `<buttons>`) to demonstrate this. I just want to confirm: would this re-ordering apply to _all_ content (i.e. not just interactive elements), though? (I am thinking of the case where, for example, a screen reader user would be navigating via their virtual cursor, following the order of content in the accessibility tree.)

From the explainers, and CSS spec, I imagine the intention is for the ordering to affect all content (not just interactive elements) - effectively the accessibility tree will come from the visual order, rather than source order.

Am I understanding this correctly?

**Comment by tabatkins** [comment]
Correct, focusable elements are just the easiest way to demonstrate this without requiring people to run a11y software themselves.

**Comment by dizhang168** [comment]
Yes, this feature is for both focus navigation and the accessibility tree (the order in which AX nodes are added to the tree).

**Comment by dizhang168** [comment]
I don't think implementation wise it will be an issue. When the grid/flex/masonry layout gets created, the [rendering-defined sibling reading flow](https://drafts.csswg.org/css-display-4/#rendering-defined-sibling-reading-flow) gets stored. So, for the AT, the additional step is just looping these reading flow items (no need to calculate on the fly). If we do need to send a signal, it should be trivial to add.

**Comment by matatk** [comment]
Thank you once again for your review request, and for your work on this. We discussed it in our calls this week, and are happy to close as satisfied.
